### ISSUE TITLE
Compatibility with upstream Makefile.gappkg

### ISSUE THREAD

**Issue by orlitzky** [semigroups/Semigroups#967]
Makefile.gappkg says,

```make
# If you bundle this file with your package, please try not to edit it,
# so that we can keep it identical across all GAP packages. Instead, if
# you find that you must edit it, please submit your changes back to
# the GAP team, so that a future version of this file can be adjusted
# to cover your usecase without modifications, thus ensuring you can
# always easily update to newer version of it.
```

but using the latest copy from GAP itself doesn't work right now. I think it may be as simple as `GNUmakefile` vs `Makefile` and `.cpp` vs `cc`. If so, how annoying would it be to rename them?

**Comment by james-d-mitchell** [comment]
Thanks for the report @orlitzky. It wouldn't be annoying to rename them at all, I wonder if @fingolfin has anything to say?

**Comment by james-d-mitchell** [comment]
Thanks @fingolfin, I'm going to close the issue as you suggest. Thanks for the report @orlitzky in any case!

**Comment by orlitzky** [comment]
No problem, it's not a big deal for me, thanks for considering it. FWIW my motivation was for packaging ~60 GAP packages on Gentoo where everything is built from source. A few packages have outdated `Makefile.gappkg` that cause compilation problems, and an easy trick to fix all of them is to simply replace the bundled `Makefile.gappkg` with the one that GAP itself ships. That worked well until I tried to package Semigroups, where the GAP version of the makefile uses a different naming convention.

Simply renaming everything did work in my case, allowing me to copy over the `Makefile.gappkg` from GAP. Looking at the commit history, I see only one important deviation: the use of `GAP_CXX` instead of `GAC` for linking. I don't know why that didn't cause an issue for me, and it's moot now :)

The only other deviation is an rpath hack that IMO is just hiding some other problem. (If a distro puts a pc file for a library into `PKG_CONFIG_PATH`, they should also put that library's libdir into `LD_LIBRARY_PATH`; it's not something for every consumer of libsemigroups to try to fix.)

tl;dr I think the only meaningful deviation from the upstream file is the `GAC` vs `GAP_CXX`. It would make diffing future versions of the Makefiles easier if they could be synchronized to that extent, but it's not a problem for me either way.

**Comment by james-d-mitchell** [comment]
> (If a distro puts a pc file for a library into `PKG_CONFIG_PATH`, they should also put that library's libdir into `LD_LIBRARY_PATH`; it's not something for every consumer of libsemigroups to try to fix.)

Sounds like this is an issue for `libsemigroups`? If so, then what would the correct fix for this be? I'm not completely sure how we'd "put that library's libdir into `LD_LIBRARY_PATH`"? Happy to make whatever changes might be needed to libsemigroups @orlitzky

**Comment by orlitzky** [comment]
> Sounds like this is an issue for `libsemigroups`? If so, then what would the correct fix for this be? I'm not completely sure how we'd "put that library's libdir into `LD_LIBRARY_PATH`"? Happy to make whatever changes might be needed to libsemigroups @orlitzky

libsemigroups is doing nothing wrong. AFAIK the only situation where the rpath hack could be helpful is if libsemigroups was installed to a nonstandard location (such as `/home/mjo/usr`) and if `PKG_CONFIG_PATH` was updated to include `/home/mjo/usr/lib/pkgconfig`, but `LD_LIBRARY_PATH` was not updated to include `/home/mjo/usr/lib`. That's up to the person who installed it (or packaged it for the distro or homebrew or whatever) and not libsemigroups itself.

Installing libraries to nonstandard locations has always been tricky. In the olden days, the `./configure` script for the Semigroups package would have done a compile/run test, linking a sample program against libsemigroups, to see if it was available. That would have worked great (well, as expected) in this example because it would have failed: even if you pass `-L/home/mjo/usr/lib` to the linker, libsemigroups isn't where ld.so expects it to be at runtime. So the test program would have crashes, and `./configure` would have complained about that, and I would have had to fix it by updating my `LD_LIBRARY_PATH` to include `/home/mjo/usr/lib`.

These days, pkg-config is supposed to make all that easier. But it opens the door to what I think happened here: if you install a library in a nonstandard location, and if the `./configure` check is using pkg-config, you can make the check "pass" by appending things to `PKG_CONFIG_PATH`, but it won't actually make the library work. You would still need to update `LD_LIBRARY_PATH` on the system with the weird configuration.

If @dimpase remembers what motivated that patch, I'm always happy to go off on a side quest.

**Comment by dimpase** [comment]
in Sage we are fine with building, installing, and using libsemigroups to a non-standard location. I don't know what's happening here, will have to look.

**Comment by orlitzky** [comment]
This may have been completely unrelated to sage but you authored https://github.com/semigroups/Semigroups/commit/dde6341b24681a7c674b99f07730d20bc552afd1 so are the best person to ask :)

**Comment by orlitzky** [comment]
sage-env is prepending `-Wl,-rpath-link,$SAGE_LOCAL/lib` to `LDFLAGS` to the same effect, so I don't think it needs https://github.com/semigroups/Semigroups/commit/dde6341b24681a7c674b99f07730d20bc552afd1

**Comment by dimpase** [comment]
This is independent of sagemath, thusly it is needed

**Comment by orlitzky** [comment]
> This is independent of sagemath, thusly it is needed

Do you remember which system needed it?

I'm wondering because setting `LD_LIBRARY_PATH` (or mangling rpath) should be the responsibility of whatever put libsemigroups in the nonstandard path to begin with. Sage is a good example: it installs libsemigroups to `$SAGE_LOCAL/lib`, but then it also makes sure to link everything with `Wl,-rpath-link,$SAGE_LOCAL/lib`. (Alternately, sage could have added `$SAGE_LOCAL/lib` to `LD_LIBRARY_PATH`).

If there's a system that puts libsemigroups in a nonstandard location but does NOT tweak the library path, I would consider that a bug on that system rather than something that Semigroups (and the other downstream consumers of libsemigroups) needs to handle.

**Comment by dimpase** [comment]
probably it was OpenBSD that needed it.

**Comment by dimpase** [comment]
I think rpath is a good idea - consider e.g. creating a python module using libsemigroups. Certainly the natural thing would be to link with rpath, as one has no control where libsemigroups.so ends up while one installs a binary wheel.

Here the situation is similar. We probably can remove the sage-env hack then.

**Comment by orlitzky** [comment]
> I think rpath is a good idea - consider e.g. creating a python module using libsemigroups. Certainly the natural thing would be to link with rpath, as one has no control where libsemigroups.so ends up while one installs a binary wheel.

I'm not sure I follow. Are you shipping libsemigroups.so inside of a python wheel (for the uninitiated: a "wheel" is just a zip file) to be installed with pip? Or shipping a cython module that links against the system libsemigroups?

What's fishy to me is that there's nothing specific about (lib)semigroups in this example. You can make the same argument about every other package that uses a library, but we're not going around adding `-rpath` to every Makefile.

> Here the situation is similar. We probably can remove the sage-env hack then.

If we remove rpath from sage's LDFLAGS, then every program built by sage and linked to something under `$SAGE_LOCAL/lib` will crash when you run it. One of `-rpath` or `LD_LIBRARY_PATH` is required to find the libraries under `$SAGE_LOCAL`. We _could_ try to add the same `-rpath` hack to the Makefiles of every sage package (rather than once in `sage-env`), but that would probably start to feel like a bad idea.

**Comment by dimpase** [comment]
I misunderstood the hack in sage-env.

But the hack in sage-env won't help if `libsemigroups.so` is, say, somewhere in your `$HOME`. This is the gift (or the curse) of pkg-config, allowing you to put `*.so` to link against anywhere in the filesystem (but you'd need rpath)

**Comment by orlitzky** [comment]
> I misunderstood the hack in sage-env.
> 
> But the hack in sage-env won't help if `libsemigroups.so` is, say, somewhere in your `$HOME`. This is the gift (or the curse) of pkg-config, allowing you to put `*.so` to link against anywhere in the filesystem (but you'd need rpath)

Yeah, this is the example I gave above. But I think if I install libsemigroups in `/home/mjo/usr/lib`, then it's my responsibility to either add that to `LD_LIBRARY_PATH` or hack the rpath of anything I link to it. Otherwise you have to change every build system on earth to support the possibility of libraries being discoverable by pkg-config but not usable when linked with `-lfoo`.

**Comment by dimpase** [comment]
yes, it's my responsibility to make sure that no matter where libsemigroup.so ends up, this GAP package links to it. Hence https://github.com/semigroups/Semigroups/commit/dde6341b24681a7c674b99f07730d20bc552afd1

Why it is a big deal for you, I don't understand.

**Comment by orlitzky** [comment]
> yes, it's my responsibility to make sure that no matter where libsemigroup.so ends up, this GAP package links to it. Hence [dde6341](https://github.com/semigroups/Semigroups/commit/dde6341b24681a7c674b99f07730d20bc552afd1)
> 
> Why it is a big deal for you, I don't understand.

It's not, I'm just trying to understand the reasoning behind it to be sure that I haven't overlooked any use cases. Thanks for taking the time to help me investigate. I'll go do something else :)

**Comment by fingolfin** [comment]
The -rpath hack is necessary on certain systems completely independent from Sage -- namely when using Semigroups with the bundled libsemigroups. Regrettably we didn't add comments next to it pointing out details, so one would have to trace to several issues and PRs to recover the details :-(.

But suffice to say: don't touch this unless you have done that work (i.e. determined all affected configurations) and can convince us that your alternative solution handles them all. If the only driver here is "purity", please don't bother (I mean this in the best possible way: I don't want to see your time wasted! Because while I agree it is desirable to fully understand the problem and have a "correct" solution, (a) in the end we need to be pragmatic and take what actually works, and (b) I am reasonably sure I was deep in the problem at one point and considered and discarded many alternatives ; while I wouldn't be surprised if someone finds a better solution, it won't be easy, and for what? I am sure there are other tasks with actual payoff that would benefit more from you skills!

**Comment by orlitzky** [comment]
I can see how it would be needed to dynamically link the bundled libsemigroups, but in that case, why use a separate variable for it and not just append it to `LDFLAGS`, leaving the build rule in `Makefile.gappkg` is untouched?

In other words, why not treat that linker flag just like you do the bundled-semigroups `CFLAGS` in e.g. 
https://github.com/semigroups/Semigroups/blob/main/GNUmakefile.in#L6C1-L7C15

Regardless, I too can be pragmatic sometimes, and am happy to drop it.

**Comment by fingolfin** [comment]
I think `LIBSEMIGROUPS_RPATH` could indeed be just appended to `KEXT_LDFLAGS` -- done in PR #976.

But that won't remove the other diffs in `Makefile.gappkg`, the most crucial of which is the switch from `GAC` to `GAP_CXX` for linking which is absolutely essential.

**Comment by orlitzky** [comment]
> I think `LIBSEMIGROUPS_RPATH` could indeed be just appended to `KEXT_LDFLAGS` -- done in PR #976.
> 
> But that won't remove the other diffs in `Makefile.gappkg`, the most crucial of which is the switch from `GAC` to `GAP_CXX` for linking which is absolutely essential.

Cool, thanks. I agree that `GAC` vs `GAP_CXX` should not be messed with, and now it's a little bit easier to see the important change in the diff.
